File: src/main/java/main/commands/prefixcommands/audio/JumpCommand.java
package main.commands.prefixcommands.audio;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import lavalink.client.player.IPlayer;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.logs.LogType;
import main.utils.json.logs.LogUtils;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.annotations.ForRemoval;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import javax.annotation.Nullable;
import javax.script.ScriptException;
import java.util.List;
import java.util.concurrent.TimeUnit;

@Deprecated @ForRemoval
public class JumpCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final Message msg = ctx.getMessage();
        final Member self = ctx.getSelfMember();
        final GuildVoiceState selfVoiceState = self.getVoiceState();
        final Member member = ctx.getMember();
        final GuildVoiceState memberVoiceState = member.getVoiceState();

        msg.replyEmbeds(doJump(selfVoiceState, memberVoiceState, ctx, null).build()).queue();
    }

    public EmbedBuilder doJump(GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState, @Nullable CommandContext ctx, @Nullable String input) {
        final var guild = memberVoiceState.getGuild();

        EmbedBuilder eb;
        if (!selfVoiceState.inVoiceChannel()) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);
            return eb;
        }

        if (!memberVoiceState.inVoiceChannel()) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL);
            return eb;
        }

        if (!memberVoiceState.getChannel().equals(selfVoiceState.getChannel())) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));
            return eb;
        }

        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(selfVoiceState.getGuild());
        final var audioPlayer = musicManager.getPlayer();
        AudioTrack track = audioPlayer.getPlayingTrack();

        if (track == null) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);
            return eb;
        }

        if (ctx != null)
            if (ctx.getArgs().isEmpty()) {
                eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JumpMessages.JUMP_MISSING_AMOUNT);
                return eb;
            } else
                return doActualJump(ctx.getGuild(), memberVoiceState.getMember().getUser(), ctx.getArgs().get(0), audioPlayer, track);
        else {
            if (input == null)
                throw new NullPointerException("Input string cannot be null");
            return doActualJump(guild, memberVoiceState.getMember().getUser(), input, audioPlayer, track);
        }
    }

    private EmbedBuilder doActualJump(Guild guild, User jumper, String input, IPlayer player, AudioTrack track) {
        long time;
        EmbedBuilder eb;
        if (GeneralUtils.stringIsInt(input))
            time = Long.parseLong(input);
        else {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JumpMessages.JUMP_INVALID_DURATION);
            return eb;
        }

        if (time <= 0) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JumpMessages.JUMP_DURATION_NEG_ZERO);
            return eb;
        }

        time = TimeUnit.SECONDS.toMillis(time);

        if (time > track.getInfo().length - player.getTrackPosition()) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JumpMessages.JUMP_DURATION_GT_TIME_LEFT);
            return eb;
        }

        player.seekTo(player.getTrackPosition() + time);
        new LogUtils().sendLog(guild, LogType.TRACK_JUMP, RobertifyLocaleMessage.JumpMessages.JUMPED_LOG, Pair.of("{user}", jumper.getAsMention()), Pair.of("{duration}", String.valueOf(time)));

        return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.JumpMessages.JUMPED, Pair.of("{duration}", String.valueOf(time)));
    }

    @Override
    public String getName() {
        return "jump";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+ GeneralUtils.listToString(getAliases()) +"`\n" +
                "Skips the song by the given number of seconds\n" +
                "\nUsage: `"+ prefix+"jump <seconds_to_jump>` *(Skips the song to a specific duration)*\n";
    }

    @Override
    public List<String> getAliases() {
        return List.of("j", "ff", "fastforward");
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/JumpCommand.kt
package main.commands.slashcommands.audio

import dev.arbjerg.lavalink.client.LavalinkPlayer
import dev.arbjerg.lavalink.client.protocol.Track
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.length
import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.utils.GeneralUtils.isInt
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.json.logs.LogType
import main.utils.json.logs.LogUtilsKt
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.JumpMessages
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.GuildVoiceState
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType
import kotlin.time.DurationUnit
import kotlin.time.toDuration

class JumpCommand : AbstractSlashCommand(
    SlashCommand(
        name = "jump",
        description = "Skip the playing song by a given number of seconds",
        options = listOf(
            CommandOption(
                type = OptionType.INTEGER,
                name = "seconds",
                description = "The number of seconds to skip in the song"
            )
        )
    )
) {

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        event.deferReply().queue()
        val memberVoiceState = event.member!!.voiceState!!
        val selfVoiceState = event.guild!!.selfMember.voiceState!!

        event.hook.sendEmbed {
            handleJump(
                selfVoiceState = selfVoiceState,
                memberVoiceState = memberVoiceState,
                input = event.getRequiredOption("seconds").asInt.toString()
            )
        }
            .setEphemeral(true)
            .queue()
    }

    private suspend fun handleJump(
        selfVoiceState: GuildVoiceState,
        memberVoiceState: GuildVoiceState,
        input: String
    ): MessageEmbed {

        val channelChecks = audioChannelChecks(memberVoiceState, selfVoiceState)
        if (channelChecks != null) return channelChecks

        val guild = selfVoiceState.guild
        val musicManager = RobertifyAudioManager[guild]
        val player = musicManager.player!!
        val track = player.track
            ?: return RobertifyEmbedUtils.embedMessage(
                guild,
                GeneralMessages.NOTHING_PLAYING
            ).build()

        return doJump(
            guild = guild,
            jumper = memberVoiceState.member.user,
            input = input,
            player = player,
            track = track
        )
    }

    private suspend fun doJump(
        guild: Guild,
        jumper: User,
        input: String,
        player: LavalinkPlayer,
        track: Track
    ): MessageEmbed {
        var time = if (input.isInt())
            input.toLong()
        else run {
            return RobertifyEmbedUtils.embedMessage(
                guild,
                JumpMessages.JUMP_INVALID_DURATION
            ).build()
        }

        if (time <= 0)
            return RobertifyEmbedUtils.embedMessage(
                guild,
                JumpMessages.JUMP_DURATION_NEG_ZERO
            ).build()

        time = time.toDuration(DurationUnit.SECONDS).inWholeMilliseconds

        if (time > track.length - player.position)
            return RobertifyEmbedUtils.embedMessage(
                guild,
                JumpMessages.JUMP_DURATION_GT_TIME_LEFT
            ).build()

        player.setPosition(player.position + time)
        LogUtilsKt(guild).sendLog(
            LogType.TRACK_JUMP,
            JumpMessages.JUMPED_LOG,
            Pair("{user}", jumper.asMention),
            Pair("{duration}", time.toDuration(DurationUnit.MILLISECONDS).inWholeSeconds.toString())
        )

        return RobertifyEmbedUtils.embedMessage(
            guild,
            JumpMessages.JUMPED,
            Pair("{duration}", time.toDuration(DurationUnit.MILLISECONDS).inWholeSeconds.toString())
        ).build()
    }
}